lib/ecto/changeset.ex

```diff
@@ -288,7 +288,7 @@ defmodule Ecto.Changeset do
                         validations: [{atom, term}],
                         filters: %{optional(atom) => term},
                         action: action,
-                        types: nil | %{atom => Ecto.Type.t}}
+                        types: nil | %{atom => Ecto.Type.t | Relation.t}}
```

Comment:
I tested this change and it doesn't solve the issue. However, in investigating (and testing) further, I think I found the right typespec for both the `:embed` and `:assoc` tuples:

```suggestion
                        types: nil | %{atom => Ecto.Type.t | {:assoc, Ecto.Association.t()} | {:embed, Ecto.Embedded.t()}}
```

I don't have any embedded schemas to test with, but I believe that this is correct. The `:assoc` tuple definitely solves my issue.